chiron/multistate.py

```diff
@@ -0,0 +1,968 @@
+        self._check_nan_energy()
+
+    def run(self, n_iterations: Optional[int] = None) -> None:
+        """
+        Execute the replica-exchange simulation.
+
+        Run the simulation for a specified number of iterations. If no number is
+        specified, it runs for the number of iterations set during the initialization
+        of the sampler.
+
+        Parameters
+        ----------
+        n_iterations : Optional[int], default=None
+            The number of iterations to run. If None, the sampler runs for the
+            number of iterations specified at initialization.
+
+        Raises
+        ------
+        RuntimeError
+            If an error occurs during the computation of energies.
+        """
+
+        # If this is the first iteration, compute and store the
+        # starting energies of the minimized/equilibrated structures.
+
+        log.info("Running simulation...")
+
+        # Initialize energies if this is the first iteration
+        if self._iteration == 0:
+            self._compute_energies()
+
+            self._reporter.write_energies(
+                energy_thermodynamic_states=self._energy_thermodynamic_states,
+                energy_neighborhoods=self._neighborhoods,
+                energy_unsampled_states=self._energy_unsampled_states,
+                iteration=self._iteration,
+            )
+
+        from openmmtools.utils import Timer
+
+        timer = Timer()
+        timer.start("Run ReplicaExchange")
+        run_initial_iteration = self._iteration
+
+        # Determine the number of iterations to run before stopping.
+        iteration_limit = (
+            min(self._iteration + n_iterations, self.number_of_iterations)
+            if n_iterations is not None
+            else self.number_of_iterations
+        )
+
+        # Main loop.
+        while not self._is_completed(iteration_limit):
+            # Increment iteration counter.
+            self._iteration += 1
+
+            log.info("-" * 80)
+            log.info(f"Iteration {self._iteration}/{iteration_limit}")
+            log.info("-" * 80)
+            timer.start("Iteration")
+
+            # Update thermodynamic states
+            self._replica_thermodynamic_states = self._mix_replicas()
+
+            # Propagate replicas.
+            self._propagate_replicas()
+
+            # Compute energies of all replicas at all states
+            self._compute_energies()
+
+            # Write iteration to storage file
+            self._report_iteration()
+
+            # TODO: Update analysis
+            # self._update_analysis()
+
+            # Update timing and progress information
+            self._update_run_progress(
+                timer=timer,
+                run_initial_iteration=run_initial_iteration,
+                iteration_limit=iteration_limit,
+            )
+
+    @with_timer("Writing iteration information to storage")
+    def _report_iteration(self):
```

Comment:
All of the original openmmtools reporter stuff will need to be overhauled. I suggest we jettison all of the reporter infrastructure and build something minimal that works for us.

Reply:
This has been addressed in PR #18